File: Sources/XDecodeCore/LoganDecoder.swift
import CommonCrypto
import Foundation

public struct LoganDecoder: LogDecoder {
    public let format = LogFormat.logan
    private let credentials: [LoganCredentials]
    // The reference Objective-C decoder zero-fills empty Key/IV strings to one AES block.
    private static let unkeyedCredentials = try! LoganCredentials(
        key: Data(repeating: 0, count: kCCKeySizeAES128),
        iv: Data(repeating: 0, count: kCCBlockSizeAES128)
    )

    public init(credentials: LoganCredentials) {
        self.credentials = [credentials]
    }

    public init(credentials: [LoganCredentials]) {
        self.credentials = credentials
    }

    public func decode(_ data: Data, sourceURL: URL) throws -> DecodedLog {
        let candidates = [Self.unkeyedCredentials] + credentials.filter {
            $0 != Self.unkeyedCredentials
        }
        var lastError: (any Error)?
        for candidate in candidates {
            do {
                return try decode(data, credentials: candidate)
            } catch {
                lastError = error
            }
        }
        guard !credentials.isEmpty else {
            throw DecodeError.missingCredentials(.logan)
        }
        if credentials.count == 1, let lastError {
            throw lastError
        }
        throw DecodeError.decryptionFailed(
            "\(credentials.count) 个匹配的 Logan Key/IV 均无法解密，密钥不匹配或日志损坏"
        )
    }

    private func decode(_ data: Data, credentials: LoganCredentials) throws -> DecodedLog {
        guard !data.isEmpty else { throw DecodeError.malformed("Logan 文件为空") }
        var cursor = 0
        var output = Data()
        var decodedFrames = 0

        while cursor < data.count {
            let marker = try data.uint8(at: cursor)
            guard marker == 0x01 else {
                throw DecodeError.malformed("Logan 帧标记无效：0x\(String(marker, radix: 16))")
            }
            cursor += 1
            let encryptedSize = Int(try data.uint32BE(at: cursor))
            cursor += 4
            guard encryptedSize > 0, encryptedSize < 10_000_000 else {
                throw DecodeError.malformed("Logan 数据块长度无效：\(encryptedSize)")
            }

            let encrypted = try data.checkedRange(cursor..<(cursor + encryptedSize))
            cursor += encryptedSize
            let isUnfinishedFinalFrame = cursor == data.count
            output.append(try decodeFrame(
                encrypted,
                credentials: credentials,
                allowsUnfinishedStream: isUnfinishedFinalFrame
            ))
            decodedFrames += 1

            if cursor < data.count {
                let delimiter = try data.uint8(at: cursor)
                guard delimiter == 0 else {
                    throw DecodeError.malformed("Logan 帧分隔符无效")
                }
                cursor += 1
            }
        }

        guard decodedFrames > 0, !output.isEmpty else { throw DecodeError.emptyOutput }
        return .complete(output)
    }

    private func decodeFrame(
        _ encrypted: Data,
        credentials: LoganCredentials,
        allowsUnfinishedStream: Bool
    ) throws -> Data {
        // Match SwiftyLoganTool first, then retain compatibility with legacy NoPadding frames.
        let options = [CCOptions(kCCOptionPKCS7Padding), CCOptions(0)]
        for option in options {
            do {
                let decrypted = try AES.decryptCBC(
                    encrypted,
                    credentials: credentials,
                    options: option
                )
                return try decodeCompressedFrame(
                    decrypted,
                    allowsUnfinishedStream: allowsUnfinishedStream && option == 0
                )
            } catch {
                continue
            }
        }

        throw DecodeError.decryptionFailed(
            "Logan Key/IV 不匹配，或 AES/压缩数据损坏"
        )
    }

    private func decodeCompressedFrame(
        _ decrypted: Data,
        allowsUnfinishedStream: Bool
    ) throws -> Data {
        let output: Data
        do {
            output = try CompressionUtilities.inflateZlibOrGzip(decrypted)
        } catch let decompressionError as DecodeError {
            guard allowsUnfinishedStream else { throw decompressionError }
            return try recoverUnfinishedFrame(decrypted)
        }

        guard !output.isEmpty else { throw DecodeError.emptyOutput }
        guard String(data: output, encoding: .utf8) != nil else {
            throw DecodeError.decodingFailed("Logan 解密内容不是有效的 UTF-8 文本")
        }
        return output
    }

    private func recoverUnfinishedFrame(_ decrypted: Data) throws -> Data {
        // CLogan can persist full AES blocks before the active gzip stream has been finalized.
        let recovered = try CompressionUtilities.inflateUnfinishedZlibOrGzip(decrypted)
        guard let lastNewline = recovered.lastIndex(of: 0x0A) else {
            throw DecodeError.decompressionFailed("Logan 未完成末帧没有完整日志行")
        }
        let completeLines = Data(recovered.prefix(through: lastNewline))
        guard !completeLines.isEmpty else { throw DecodeError.emptyOutput }
        guard String(data: completeLines, encoding: .utf8) != nil else {
            throw DecodeError.decodingFailed("Logan 未完成末帧不是有效的 UTF-8 文本")
        }
        return completeLines
    }
}

private enum AES {
    static func decryptCBC(
        _ data: Data,
        credentials: LoganCredentials,
        options: CCOptions
    ) throws -> Data {
        guard data.count.isMultiple(of: kCCBlockSizeAES128) else {
            throw DecodeError.decryptionFailed("AES-CBC 数据长度不是 16 的倍数")
        }
        let outputCapacity = data.count + kCCBlockSizeAES128
        var output = Data(count: outputCapacity)
        var moved = 0

        let status = output.withUnsafeMutableBytes { outputBuffer in
            data.withUnsafeBytes { inputBuffer in
                credentials.key.withUnsafeBytes { keyBuffer in
                    credentials.iv.withUnsafeBytes { ivBuffer in
                        CCCrypt(
                            CCOperation(kCCDecrypt),
                            CCAlgorithm(kCCAlgorithmAES),
                            options,
                            keyBuffer.baseAddress,
                            credentials.key.count,
                            ivBuffer.baseAddress,
                            inputBuffer.baseAddress,
                            data.count,
                            outputBuffer.baseAddress,
                            outputCapacity,
                            &moved
                        )
                    }
                }
            }
        }

        guard status == kCCSuccess else {
            throw DecodeError.decryptionFailed("CommonCrypto 状态码 \(status)")
        }
        output.removeSubrange(moved..<output.count)
        return output
    }
}

File: Windows/src/XDecode.Core/LoganDecoder.cs
using System.Security.Cryptography;
using System.Text;

namespace XDecode.Core;

public sealed class LoganDecoder : ILogDecoder
{
    private static readonly LoganCredentials Unkeyed = new(new byte[16], new byte[16]);
    private static readonly UTF8Encoding StrictUtf8 = new(false, true);
    private readonly LoganCredentials[] _credentials;
    public LogFormat Format => LogFormat.Logan;

    public LoganDecoder(LoganCredentials credentials) : this([credentials]) { }
    public LoganDecoder(IEnumerable<LoganCredentials> credentials) => _credentials = credentials.ToArray();

    public DecodedLog Decode(ReadOnlyMemory<byte> data, string sourcePath)
    {
        var candidates = new[] { Unkeyed }.Concat(_credentials.Where(value => !CredentialsEqual(value, Unkeyed)));
        Exception? lastError = null;
        foreach (var candidate in candidates)
        {
            try { return DecodeWithCredentials(data.Span, candidate); }
            catch (DecodeException exception) when (exception.Kind == DecodeErrorKind.OutputLimitExceeded) { throw; }
            catch (Exception exception) { lastError = exception; }
        }
        if (_credentials.Length == 0) throw DecodeException.MissingCredentials(LogFormat.Logan);
        if (_credentials.Length == 1 && lastError is not null) throw lastError;
        throw DecodeException.DecryptionFailed($"{_credentials.Length} 个匹配的 Logan Key/IV 均无法解密，密钥不匹配或日志损坏");
    }

    private static DecodedLog DecodeWithCredentials(ReadOnlySpan<byte> data, LoganCredentials credentials)
    {
        if (data.IsEmpty) throw DecodeException.Malformed("Logan 文件为空");
        using var output = new MemoryStream();
        var cursor = 0;
        var frames = 0;
        while (cursor < data.Length)
        {
            var marker = data.UInt8(cursor++);
            if (marker != 0x01) throw DecodeException.Malformed($"Logan 帧标记无效：0x{marker:x}");
            var encryptedSize = checked((int)data.UInt32BE(cursor));
            cursor += 4;
            if (encryptedSize is <= 0 or >= 10_000_000)
                throw DecodeException.Malformed($"Logan 数据块长度无效：{encryptedSize}");
            var encrypted = data.CheckedRange(cursor, encryptedSize);
            cursor += encryptedSize;
            var unfinishedFinalFrame = cursor == data.Length;
            var decoded = DecodeFrame(
                encrypted, credentials, unfinishedFinalFrame,
                checked(DecodeLimits.MaximumDecompressedOutputSize - (int)output.Length));
            DecodeLimits.ValidateOutputSize(output.Length, decoded.Length);
            output.Write(decoded);
            frames++;
            if (cursor < data.Length && data.UInt8(cursor++) != 0)
                throw DecodeException.Malformed("Logan 帧分隔符无效");
        }
        if (frames == 0 || output.Length == 0) throw DecodeException.EmptyOutput();
        return DecodedLog.Complete(output.ToArray());
    }

    private static byte[] DecodeFrame(
        ReadOnlySpan<byte> encrypted,
        LoganCredentials credentials,
        bool allowsUnfinishedStream,
        int maximumOutputSize)
    {
        foreach (var padding in new[] { PaddingMode.PKCS7, PaddingMode.None })
        {
            try
            {
                var decrypted = DecryptAesCbc(encrypted, credentials, padding);
                return DecodeCompressedFrame(
                    decrypted, allowsUnfinishedStream && padding == PaddingMode.None, maximumOutputSize);
            }
            catch (DecodeException exception) when (exception.Kind == DecodeErrorKind.OutputLimitExceeded) { throw; }
            catch { }
        }
        throw DecodeException.DecryptionFailed("Logan Key/IV 不匹配，或 AES/压缩数据损坏");
    }

    private static byte[] DecodeCompressedFrame(byte[] decrypted, bool allowsUnfinishedStream, int maximumOutputSize)
    {
        try
        {
            var output = CompressionUtilities.InflateZlibOrGzip(decrypted, maximumOutputSize);
            ValidateUtf8(output, "Logan 解密内容不是有效的 UTF-8 文本");
            return output;
        }
        catch (DecodeException exception) when (
            allowsUnfinishedStream && exception.Kind != DecodeErrorKind.OutputLimitExceeded)
        {
            var recovered = CompressionUtilities.InflateUnfinishedZlibOrGzip(decrypted, maximumOutputSize);
            var lastNewline = Array.LastIndexOf(recovered, (byte)'\n');
            if (lastNewline < 0)
                throw DecodeException.DecompressionFailed("Logan 未完成末帧没有完整日志行");
            var completeLines = recovered[..(lastNewline + 1)];
            ValidateUtf8(completeLines, "Logan 未完成末帧不是有效的 UTF-8 文本");
            return completeLines;
        }
    }

    private static byte[] DecryptAesCbc(
        ReadOnlySpan<byte> data, LoganCredentials credentials, PaddingMode padding)
    {
        if (data.Length % 16 != 0)
            throw DecodeException.DecryptionFailed("AES-CBC 数据长度不是 16 的倍数");
        try
        {
            using var aes = Aes.Create();
            aes.Mode = CipherMode.CBC;
            aes.Padding = padding;
            aes.Key = credentials.Key;
            aes.IV = credentials.IV;
            using var decryptor = aes.CreateDecryptor();
            return decryptor.TransformFinalBlock(data.ToArray(), 0, data.Length);
        }
        catch (CryptographicException exception)
        {
            throw DecodeException.DecryptionFailed(exception.Message, exception);
        }
    }

    private static void ValidateUtf8(byte[] data, string message)
    {
        if (data.Length == 0) throw DecodeException.EmptyOutput();
        try { _ = StrictUtf8.GetString(data); }
        catch (DecoderFallbackException) { throw DecodeException.DecodingFailed(message); }
    }

    private static bool CredentialsEqual(LoganCredentials left, LoganCredentials right) =>
        left.Key.AsSpan().SequenceEqual(right.Key) && left.IV.AsSpan().SequenceEqual(right.IV);
}
